mqttv5/src/main/scala/org/apache/pekko/stream/connectors/mqttv5/impl/MqttFlowStage.scala

```diff
@@ -0,0 +1,516 @@
+                      token.getReasonCodes.distinct.sorted.mkString(";"),
+                      token.getResponseProperties.getReasonString,
+                      s"${ex.getClass.getSimpleName} - ${ex.getMessage}"
+                    )
+                }
+              )
+            }
+
+          case other =>
+            // unexpected reason code received (if enhanced authentication is used)
+            // OR
+            // unexpected AUTH packet received (if no or simple authentication is used)
+            log.warning(
+              "Client [{}] with authentication type [{}] received an unexpected AUTH packet with [code={},reason={}]",
+              connectionSettings.clientId,
+              other.getClass.getSimpleName.replaceAll("[^a-zA-Z0-9]", "").toLowerCase,
+              reasonCode,
+              properties.getReasonString
+            )
+        }
+      }
+
+      override def connectComplete(reconnect: Boolean, serverURI: String): Unit = {
+        log.debug(
+          "Connection completed for client [{}] with [reconnect={},serverURI={}]",
+          connectionSettings.clientId,
+          reconnect,
+          serverURI
+        )
+        pendingMsg.foreach { msg =>
+          log.debug("Client [{}] sending pending message to broker [{}]", connectionSettings.clientId, serverURI)
+          publishPending(msg)
+          pendingMsg = None
+        }
+        if (reconnect && !hasBeenPulled(in)) pull(in)
+      }
+    }
+  )
+
+  override def onPush(): Unit = {
+    val msg = grab(in)
+    try {
+      publishPending(msg)
+    } catch {
+      case _: MqttException if connectionSettings.automaticReconnect => pendingMsg = Some(msg)
+      case NonFatal(e)                                               => throw e
+    }
+  }
+
+  override def onUpstreamFinish(): Unit = {
+    setKeepGoing(true)
+    if (queue.isEmpty && unackedMessages.get() == 0) super.onUpstreamFinish()
+  }
+
+  override def onUpstreamFailure(ex: Throwable): Unit = {
+    setKeepGoing(true)
+    if (queue.isEmpty && unackedMessages.get() == 0) super.onUpstreamFailure(ex)
+  }
+
+  override def onPull(): Unit =
+    if (queue.nonEmpty) {
+      pushDownstream(queue.dequeue())
+      if (unackedMessages.get() == 0 && isClosed(in)) completeStage()
+    }
+
+  override def onDownstreamFinish(cause: Throwable): Unit = {
+    setKeepGoing(true)
+    if (unackedMessages.get() == 0) super.onDownstreamFinish(cause)
+  }
+
+  setHandlers(in, out, this)
+
+  def publishToMqtt(msg: MqttMessage): IMqttToken = {
+    val pahoMsg = new PahoMqttMessage(msg.payload.toArray)
+    pahoMsg.setQos(msg.qos.getOrElse(defaultQoS).value)
+    pahoMsg.setRetained(msg.retained)
+
+    mqttClient.publish(
+      msg.topic,
+      pahoMsg,
+      msg,
+      /* callback */ new MqttActionListener {
+        def onSuccess(token: IMqttToken): Unit = {
+          token.getReasonCodes.toList.filter(_ >= MqttReturnCode.RETURN_CODE_UNSPECIFIED_ERROR).distinct match {
```

Comment:
[info] As with the subscription above, the return codes are checked and the stage is failed if something went wrong.